vignettes/sir_models.Rmd

```diff
@@ -0,0 +1,680 @@
+
+```{r}
+index <- function(info) {
+  list(run = c(info$index$S, info$index$D), state = c(info$index$S, info$index$I, info$index$R, info$index$D))
+}
+
+index(sir_death$info())
+```
+
+Which sends the comparison function the indices listed under `run`, in this case the number of susceptibles, and saves any compartments listed under `state` if the history is kept. In the `compare()` function the index of `state` now corresponds to `index$run`. The missing data must be checked for, and contribute 0 to the likelihood if found:
+
+```{r}
+combined_compare <- function(state, prev_state, observed, pars = NULL) {
+  if (is.na(observed$cases) && is.na(observed$deaths)) {
+    return(NULL)
+  }
+  if (is.null(pars$compare$exp_noise)) {
+    exp_noise <- 1e6
+  } else {
+    exp_noise <- pars$compare$exp_noise
+  }
+  
+  ll_pois <- function(obs, model, exp_noise) {
+    if (is.na(obs)) {
+      ll_obs <- numeric(length(model))
+    } else {
+      lambda <- model +
+        rexp(n = length(model), rate = exp_noise)
+      ll_obs <- dpois(x = obs, lambda = lambda, log = TRUE)
+    }
+    ll_obs
+  }
+  
+  ll_cases <- ll_pois(observed$cases, 
+                      prev_state[1, , drop = TRUE] - state[1, , drop = TRUE],
+                      exp_noise)
+  ll_deaths <- ll_pois(observed$deaths, 
+                      state[2, , drop = TRUE] - prev_state[2, , drop = TRUE],
+                      exp_noise)
+  
+  ll_cases + ll_deaths
+}
+```
+
+This can be used to set up a particle filter as before:
+
+```{r}
+n_particles <- 100
+filter <- particle_filter$new(data = combined_data,
+                              model = gen_death,
+                              n_particles = n_particles,
+                              compare = combined_compare,
+                              index = index,
+                              n_threads = 4L,
+                              seed = 1L)
+filter$run(save_history = FALSE, pars = list(dt = dt))
+```
+
+and this can be used to run the `pmcmc()` method, as above.
+
+## Running predictions
+
+Going back to the simple SIR model above, we can easily continue running the fitted model forward in time using the posterior estimated by `pmcmc()` to create a forecast past the end of the data:
+
+```{r fig.height=5, fig.width=7}
+forecast <- predict(processed_chains, 
+                    steps = seq(400, 800, 4), 
+                    n_threads = 4L, 
+                    prepend_trajectories = TRUE,
+                    seed = processed_chains$predict$seed)
+
+# Sample 10 particles
+mini_forecast <- forecast$state[,sample.int(dim(forecast$state)[2], size = 10),]
+time <- mini_forecast[1,1,-1]
```

Comment:
Looks like linting is not on here by default. Can you look over the style and fix the bulk of these please? I can do this if you'd rather not

Reply:
@richfitz linting the vignettes is a mystery to me unfortunately. If I run lintr::lint_package() they are ignored, same with lint_dir. You can run lint on the Rmd files directly and it works, but I don't know how to do this just from the .lintr file.

https://github.com/jimhester/lintr/blob/master/NEWS.md suggests this should run by default as of v2.0.1, so I'm not sure what's going on

Reply:
Also I am unable to/incorrectly turning off linting of the odin code in CodeFactor